### PR TITLE
ADBDEV-3247 Fix "Failed to kill processes for segment" error of gpstop

### PR DESCRIPTION
**Pull request by dnskvlnk**
Geenplum python scrips try to parse system command’s STDOUT and STDERR in english and fails if locale is different from en_US

For example, /usr/lib/gpdb/lib/python/gppylib/commands/pg.py contains class ReadPostmasterTempFile with method getResults, which try to find message "No such file or directory" and fails when locale non english. This method used in gpstop in _stop_master function and on Alt and Astra gpstop prints this error:

```
20221129:16:38:21:003241 gpstop:oan-astra1:gpadmin-[INFO]:-Attempting forceful termination of any leftover master process
20221129:16:38:21:003241 gpstop:oan-astra1:gpadmin-[INFO]:-Terminating processes for segment /data1/master/gpseg-1
20221129:16:38:21:003241 gpstop:oan-astra1:gpadmin-[ERROR]:-Failed to kill processes for segment /data1/master/gpseg-1: ([Errno 3] No such process)
20221129:16:38:21:003305 gpstop:oan-astra1:gpadmin-[INFO]:-Starting gpstop with args: -a -m -f -d /data1/master/gpseg-1 -B 64 -l /home/gpadmin/gpAdminLogs
```
On Centos with locale en_US there is no error in gpstop invocation:

```
20221128:12:19:07:016251 gpstop:oan-adb-pg-hba:gpadmin-[INFO]:-Attempting forceful termination of any leftover master process
20221128:12:19:07:016251 gpstop:oan-adb-pg-hba:gpadmin-[INFO]:-Terminating processes for segment /data1/master/gpseg-1
20221128:12:19:07:016318 gpstop:oan-adb-pg-hba:gpadmin-[INFO]:-Starting gpstop with args: -a -m -f -d /data1/master/gpseg-1 -B 64 -l /home/gpadmin/gpAdminLogs
```
Geenplum python scrips try to parse system command’s STDOUT and STDERR in English and may fail if locale is different from en_US. This patch forced it to check the return code rather then a message itself.

Also, this patch reinstalls glibc-common in docker container. This is necessary to get langpacks in docker because docker images don't contain them.

### CODE REVIEW

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 1]
```diff
@@ -3962,3 +3962,11 @@ def impl(context):
     cmd =  "sudo mv -f /tmp/hosts_orig /etc/hosts; rm -f /tmp/clusterConfigFile-1; rm -f /tmp/hostfile--1"
     context.execute_steps(u'''Then the user runs command "%s"''' % cmd)
 
+@given(u'"LANG" is different from English')
+def step_impl(context):
+	os.environ['LANG'] = 'ru_RU.utf8'
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: what if ru_RU.utf8 not in available locales?
for example
```sh
LANG=qwe ls -lan diskquota*
ls: cannot access 'diskquota*': No such file or directory
LENG=qwe ls -lan diskquota*
ls: невозможно получить доступ к 'diskquota*': Нет такого файла или каталога
```

Comment (Stolb27) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: It seems that existing check doesn't validate the problem.

1. I've reverted changes for gpMgmt/bin/gppylib/commands/base.py
    ```
    git checkout adb-6.x-dev -- gpMgmt/bin/gppylib/commands/base.py
    ```
3. Build Docker container
    ```
    docker build -f arenadata/Dockerfile .
    ```
5. Run gpstop tests inside
    ```
    export IMAGE=sha256:<hash>
    bash arenadata/scripts/run_behave_tests.bash gpstop
    ```

Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: `LANG` env variable didn't affect the language, so I switched back to `LC_ALL`. Also, the error is printed to stdout, so the test was always successful, I fixed that too

Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: I also added Russian locale generation, so there's shouldn't be a problem in a missing locale

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 2]
```diff
@@ -41,7 +41,7 @@
 from gppylib.commands.base import Command, REMOTE
 from gppylib import pgconf
 
-
+default_locale = os.environ.get('LANG')
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Why `LANG` here?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 3]
```diff
@@ -3962,3 +3962,12 @@ def impl(context):
...
 
+@given(u'"LC_ALL" is different from English')
+def step_impl(context):
+	os.system('sudo locale-gen ru_RU.UTF-8 > /dev/null')
+	os.environ['LC_ALL'] = 'ru_RU.utf8'
+
+@then('gpstop should not print "Failed to kill processes for segment"')
+def impl(context):
+	check_string_not_present_stdout(context, 'Failed to kill processes for segment')
+	os.environ['LC_ALL'] = default_locale
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: While `LC_ALL` here?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 2 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Thanks, it supposed to be LC_ALL

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 3 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Fixed

**Comment by RekGRpth** [comment]
please, update branch (button in right bottom corner)

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 4]
```diff
@@ -41,7 +41,7 @@
 from gppylib.commands.base import Command, REMOTE
 from gppylib import pgconf
 
-
+default_locale = os.environ.get('LC_ALL')
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: test broken, I suggest such solution
```py
default_locale = os.environ.get('LC_ALL') or 'C'
```

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 5]
```diff
@@ -3995,3 +3995,12 @@ def impl(context):
...
 
+@given(u'"LC_ALL" is different from English')
+def step_impl(context):
+	os.system('sudo locale-gen ru_RU.UTF-8 > /dev/null')
+	os.environ['LC_ALL'] = 'ru_RU.utf8'
+
+@then('gpstop should not print "Failed to kill processes for segment"')
+def impl(context):
+	check_string_not_present_stdout(context, 'Failed to kill processes for segment')
+	os.environ['LC_ALL'] = default_locale if default_locale else 'C.utf8'
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: I didn't understand, why not simple 'C'?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 4 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Fixed

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 6]
```diff
@@ -3995,3 +3994,12 @@ def impl(context):
...
 
+@given('"LC_ALL" is different from English')
+def step_impl(context):
+	os.system('sudo locale-gen ru_RU.UTF-8 > /dev/null')
+	os.environ['LC_ALL'] = 'ru_RU.utf8'
+
+@then('gpstop should not print "Failed to kill processes for segment"')
+def impl(context):
+	check_string_not_present_stdout(context, 'Failed to kill processes for segment')
+	del os.environ['LC_ALL']
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: what if `os.environ['LC_ALL'] is not None` before setting it to ru_RU.utf8?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 7]
```diff
@@ -3995,3 +3995,15 @@ def impl(context):
     cmd =  "sudo mv -f /tmp/hosts_orig /etc/hosts; rm -f /tmp/clusterConfigFile-1; rm -f /tmp/hostfile--1"
     context.execute_steps(u'''Then the user runs command "%s"''' % cmd)
 
+@given('"LC_ALL" is different from English')
+def step_impl(context):
+	os.system('sudo locale-gen ru_RU.UTF-8 > /dev/null')
+	os.environ['LC_ALL'] = 'ru_RU.utf8'
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:4007]: may be add before this row something like
```python
default_locale = os.environ.get('LC_ALL')
```
to ensure we save actual var instead one when file was imported?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 8]
```diff
@@ -3995,3 +3995,15 @@ def impl(context):
...
+
+@then('gpstop should not print "Failed to kill processes for segment"')
+def impl(context):
+	check_string_not_present_stdout(context, 'Failed to kill processes for segment')
+	if default_locale != None:
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: on python one usually writes
```python
if default_locale is not None:
```

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 6 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: fixed

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 7 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:4007]: fixed

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 8 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: fixed

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 1 above)
Comment (RekGRpth) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: I try to reproduce error without patch as describe above, but got
```sh
sudo: locale-gen: command not found
1 feature passed, 0 failed, 23 skipped
2 scenarios passed, 0 failed, 539 skipped
12 steps passed, 0 failed, 7056 skipped, 0 undefined
Took 0m8.583s

real	0m15.825s
user	0m1.906s
sys	0m0.471s
+ status=0
+ docker-compose -p gpstop_demo -f arenadata/docker-compose.yaml --env-file arenadata/.env down -v
Stopping gpstop_demo_sdw3_1 ... done
Stopping gpstop_demo_sdw1_1 ... done
Stopping gpstop_demo_sdw2_1 ... done
Stopping gpstop_demo_cdw_1  ... done
Removing gpstop_demo_sdw3_1 ... done
Removing gpstop_demo_sdw1_1 ... done
Removing gpstop_demo_sdw2_1 ... done
Removing gpstop_demo_cdw_1  ... done
Removing network gpstop_demo_default
+ [[ 0 > 0 ]]
+ exit 0
+ exits=0
+ [[ 0 > 0 ]]
```
is `locale-gen` command Ubuntu-specific while docker image based on Centos?

**Review by RekGRpth — DISMISSED** [review]

**Review by red1452 — DISMISSED** [review]

File: arenadata/Dockerfile [hunk 9]
```diff
@@ -3,6 +3,10 @@ FROM centos:centos7 as base
 ARG sigar=https://downloads.adsw.io/ADB/6.22.0_arenadata38/centos/7/community/x86_64/sigar-1.6.5-1056.git2932df5.el7.x86_64.rpm
 ARG sigar_headers=http://downloads.adsw.io/ADB/6.22.0_arenadata38/centos/7/community/x86_64/sigar-headers-1.6.5-1056.git2932df5.el7.x86_64.rpm
 
+# Download langpacks
+RUN sed -i 's/\(override_install_langs*\)/# \1/' /etc/yum.conf
+RUN yum -y reinstall glibc-common
```
Comment (Stolb27) [inline arenadata/Dockerfile]: Please describe why we need to reinstall this package, combine steps to one (each Dockerfile command including RUN create a new layer), clean cache after reinstall.

File: gpMgmt/test/behave/mgmt_utils/gpstop.feature [hunk 10]
```diff
@@ -38,3 +38,10 @@ Feature: gpstop behave tests
           And gpstop should print "invalid_host is unreachable.Skipping cleaning shared memory." to stdout
           And gpstop should return a return code of 0
           And the standby host is made reachable
+
+	@demo_cluster
+    Scenario: gpstop exits without a warning
```
Comment (Stolb27) [inline gpMgmt/test/behave/mgmt_utils/gpstop.feature]: Provide more verbose scenery description. There are may be logged dozens of different warning during cluster stop, but you check some exact.

**Review by Stolb27 — COMMENTED** [review]
@dnskvlnk, provide a valid commit message. @RekGRpth, why you didn't check this moment?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 1 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Docker image building now involves glibc-common reinstallation which contains all missing locales, therefore locale-gen is not needed anymore

File: arenadata/Dockerfile (hunk 9 above)
Comment (dnskvlnk) [reply arenadata/Dockerfile]: Added description and combined into one command

File: gpMgmt/test/behave/mgmt_utils/gpstop.feature (hunk 10 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/gpstop.feature]: Added more verbose description

**Comment by RekGRpth** [comment]
Why not just backport https://github.com/greenplum-db/gpdb/pull/13231?

**Comment by Stolb27** [comment]
> Why not just backport [greenplum-db#13231](https://github.com/greenplum-db/gpdb/pull/13231)?

Thank you, it's quite strange but better solution to be consistent with newer version. But we need to recheck that different negative values for pid are never used.

@dnskvlnk, ^^

**Comment by RekGRpth** [comment]
> it's quite strange but better solution

This solution immediately came to my mind when I heard that they check the string description of the error.

**Comment by dnskvlnk** [comment]
> > it's quite strange but better solution
> 
> This solution immediately came to my mind when I heard that they check the string description of the error.

Yes, I thought about it too, but decided to check stderr because cat always returns 1 in a case when it fails, so it's hard to tell what exactly happened

File: arenadata/Dockerfile [hunk 11]
```diff
@@ -3,6 +3,12 @@ FROM centos:centos7 as base
 ARG sigar=https://downloads.adsw.io/ADB/6.22.0_arenadata38/centos/7/community/x86_64/sigar-1.6.5-1056.git2932df5.el7.x86_64.rpm
 ARG sigar_headers=http://downloads.adsw.io/ADB/6.22.0_arenadata38/centos/7/community/x86_64/sigar-headers-1.6.5-1056.git2932df5.el7.x86_64.rpm
 
+# Reinstall glibc-common. This is necessary to get langpacks in docker
+# because docker images don't contain them.
+RUN sed -i 's/\(override_install_langs*\)/# \1/' /etc/yum.conf && \
+    yum -y reinstall glibc-common && \
```
Comment (RekGRpth) [inline arenadata/Dockerfile:9]: Can we only do this in test images?

Comment (dnskvlnk) [reply arenadata/Dockerfile:9]: Do you mean something like this?
```
FROM base as test
RUN sed -i 's/\(override_install_langs*\)/# \1/' /etc/yum.conf && \
    yum -y reinstall glibc-common && \
	yum clean all
COPY --from=code /home/gpadmin/gpdb_src gpdb_src
COPY --from=build /home/gpadmin/bin_gpdb /home/gpadmin/bin_gpdb
```

Comment (RekGRpth) [reply arenadata/Dockerfile:9]: Something like that.

Comment (dnskvlnk) [reply arenadata/Dockerfile:9]: Done

File: arenadata/Dockerfile [hunk 12]
```diff
@@ -75,6 +75,11 @@ COPY . gpdb_src
 RUN rm -rf gpdb_src/.git/
 
 FROM base as test
+# Reinstall glibc-common. This is necessary to get langpacks in docker
```
Comment (RekGRpth) [inline arenadata/Dockerfile]: some notice for [lines](https://github.com/arenadata/gpdb/pull/481/files#diff-148eeb204cd18bca72d96990f7c3842fedbc9f03051f2e9758ef257c03547f45R37-R39): https://stackoverflow.com/a/27504514/4139676

Comment (dnskvlnk) [reply arenadata/Dockerfile]: I don't understand how this is connected to my patch

**Comment by RekGRpth** [comment]
Could You combine all the commits into two: the first one is a pure upstream cherry-pick, and the second one is Your tests, with appropriate descriptions?

**Comment by dnskvlnk** [comment]
> Could You combine all the commits into two: the first one is a pure upstream cherry-pick, and the second one is Your tests, with appropriate descriptions?

Done

**Comment by RekGRpth** [comment]
> > Could You combine all the commits into two: the first one is a pure upstream cherry-pick, and the second one is Your tests, with appropriate descriptions?
> 
> Done

IMHO, pure cherry-pick should contain `authored` and `committed`.

File: arenadata/Dockerfile (hunk 11 above)
Comment (Stolb27) [reply arenadata/Dockerfile:9]: @RekGRpth, @dnskvlnk what motivation to separate this step from another yum manipulation?

**Comment by Stolb27** [comment]
@dnskvlnk, please pay attention to this patch design:

1. As was mentioned by @RekGRpth, it's very bad style to copy somebody changes with mentioning real author. Moreover, your patch contain useless whitespace changes. Use `co-authored-by:` [tag](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) in the commit massage or in your case `git cherry-pick 240080b953ba8a427450165b1a30c9d4bde06b91` yo cover both problems. 
2. Move back or motivate changes that move yum command to test image. I think that all yum commands must be combined if there are no another logic.
3. Save split on the two commit (cherry picked fix and tests)
4. Update PR description to be consistent with real changes:
    > This patch forces 'C' locale for all subprocesses to avoid this problem

File: arenadata/Dockerfile (hunk 11 above)
Comment (dnskvlnk) [reply arenadata/Dockerfile:9]: Changes have been reverted

**Comment by RekGRpth** [comment]
I don't understand why there are still three commits here?

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**